Modules/arraymodule.c

```diff
@@ -2635,6 +2650,10 @@ array_subscr(arrayobject* self, PyObject* item)
 static int
 array_ass_subscr_lock_held(arrayobject* self, PyObject* item, PyObject* value)
 {
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(self);
+    if (value != NULL) {
+        _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(value);
+    }
```

Comment:
Use pre-processor guard instead of runtime guards for these: `Py_DEBUG`.

Reply:
The whole if block gets optimized out when `Py_DEBUG` is not defined, but having the `#ifdef` doesn't hurt either.